### ISSUE TITLE
Uncaught TypeError: Cannot read property 'mainModule' of undefined

### ISSUE THREAD

**Issue by UziTech** [morassman/process-palette#47]
I get this error when tree-view is not enabled.

It might be better to check for `.tree-view` element

**Comment by UziTech** [comment]
Looks like the problem line is at:
https://github.com/morassman/process-palette/blob/master/lib/controllers/config-controller.coffee#L97

**Comment by morassman** [comment]
I've never disabled my tree view before, but it makes sense that it would break on that line if it is disabled. I'll release a fix for this soon.

**Comment by morassman** [comment]
What is the name of the package that you are using? I'd like to test with it installed.

**Comment by UziTech** [comment]
It is just a fork of the tree-view package with a few updates

https://github.com/UziTech/tree-view

you can install it with the github url

**Comment by morassman** [comment]
Ok thanks. That makes sense now. I thought that the fork was published as a new package with its own name.

**Comment by UziTech** [comment]
The forked package does have a different name since Atom won't let core packages be uninstalled. In this case the name is `tree-view-master`

**Comment by morassman** [comment]
I've cloned your repo. I came across `tree-view-master` and thought that that might be it.

I must be honest with you. I'm not super fond of the way this issue was addressed in your pull request. I try to adhere to [MVC](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) as much as I can. One should be able to run an application's business logic without the UI. This is especially true when running tests. I know with web technologies the lines can become blurry, but with a good design it shouldn't be the case.

Let's look at the one change from the pull request:
```
selected = document.querySelectorAll(".tree-view .selected > .list-item > .name, .tree-view .selected > .name")
 +    @runProcessWithFile(selected[0].dataset.path);
```

This has a strong dependency on the UI. Not only that, but it's also sensitive to changes in the selectors. One small change and it won't work anymore. The same _can_ be said about the module's API, but APIs are typically less likely to change. If Atom used a statically typed language then the expectation would be to access the module through its API, which I consider is good practice. Just because the language being used is dynamically typed doesn't mean that one should abandon that practice.

I also think this is a bit of a corner case. I can't imagine that forking one's own version of the tree-view is something that is common. I would rather integrate with the core tree-view package be default, but then add the ability for other packages to integrate with process-palette as apposed to making sure that process-palette can integrate with all other packages from its point of view.

I'd appreciate your input on this, but thus far this is where I stand with regards to this.

**Comment by UziTech** [comment]
> This has a strong dependency on the UI. Not only that, but it's also sensitive to changes in the selectors. 

Adding a menu item is already dependent on selectors. You already specify `.tree-view` when adding the item.

```coffeescript
@ctxMenuDisposables.add(atom.contextMenu.add({'.tree-view': [root]}));
```

This is the way APIs work in a DOM based environment. I think Atom knows that if they change selectors it would be the same as changing the API.

Selectors are a great tool for this type of API, and not using them would be like using a jackhammer to pound in a nail. It would be a lot of work when you could just use a hammer.

**Comment by UziTech** [comment]
If you look at other packages that get selected files from the tree-view packages you will see that they use a similar approach.

https://github.com/icetee/remote-ftp/blob/1c261aec440a923628303342329e3c757a4ffbcd/lib/client.js#L212
https://github.com/akonwi/git-plus/blob/master/lib/context-menu.coffee#L3
https://github.com/UziTech/context-git/blob/master/lib/helper.js#L45

**Comment by morassman** [comment]
> Adding a menu item is already dependent on selectors. You already specify .tree-view when adding the item.

Yes, but that is for the context menu, which is part of the UI. It is therefore still within the realm of the UI. It's the concept of "separation of concerns" that's important to me : Model-View-Controller.

What I'm trying to get at is that from process-palette's point of view I do not want to add support for integrating with every other package that shows files. If using the selector to get the path is how the tree-view package _expects_ one to get the path _and_ it can ensure that it will not change, then I'll be fine with the solution. If forks of the tree-view or any other package happens to use the same selectors then that's great, but I don't want to have to cater for multiple packages, APIs, selectors, etc.

> If you look at other packages that get selected files from the tree-view packages you will see that they use a similar approach.

Thanks for the links. It motivates your point and it helps to see how this kind of thing gets done by others.

**Comment by UziTech** [comment]
I understand, and I agree that process-palette shouldn't try to accommodate all other packages that show files (IMHO adding support for the `tabs` package would be nice though) but I do believe this is the way `tree-view` intended on other packages to get selected files.

**Comment by hg42** [comment]
I agree with @morassman...

I am not familar with the tree-view, and currently very busy with other things, but a quick search gave me this:

```
  # Public: Return an array of paths from all selected items
  #
  # Example: @selectedPaths()
  # => ['selected/path/one', 'selected/path/two', 'selected/path/three']
  # Returns Array of selected item paths
  selectedPaths: ->
    entry.getPath() for entry in @getSelectedEntries()
```
in [tree-view.coffee](https://github.com/atom/tree-view/blob/master/lib/tree-view.coffee)

isn't this exactly what we need?
(I apologize if I'm missing something)

**Comment by UziTech** [comment]
Yes that will get the selected paths for the tree-view package, but I don't think they meant for that to be a public API

**Comment by UziTech** [comment]
The only reason for tree-view to put the path of files in the `dataset` of a file element is for other packages to be able to get that path.

**Comment by morassman** [comment]
When you look at my implementation you'll see that I do access tree-view.coffee. Instead of the array, selectedPaths, I use the singular, selectedPath. @UziTech makes a good point though and I'm starting to see the logic behind it.

I've integrated with the tabs package before actually. Before Atom had the ability to dock tabs by dragging them I published [tabasco](https://github.com/morassman/tabasco) that added that functionality. It's deprecated now though due to the built-in behaviour. The point being :) integrating with the tabs package is not out of the question. I'll look into it some time.

In the mean time I will merge in the changes made by @UziTech.
